### ISSUE TITLE
[Feature Request] Hot-plug vCPUs

### ISSUE THREAD

**Issue by jeromegn** [firecracker-microvm/firecracker#2609]
# Feature Request

We'd like to be able to add (and remove) vCPUs from running firecracker microvms. It appears to be possible to do that with KVM. Examples I've seen define how many vCPUs at most a VM might use and then the actual number it will be using at boot. Then you can add vCPUs via `virsh`.

https://www.unixarena.com/2015/12/linux-kvm-how-to-add-remove-vcpu-to-guest-on-fly.html/

This would allow use to add a "burst" feature when CPU usage spikes.

## Describe the desired solution

An API to modify a running microvm's vCPUs count.

This should notify the guest VM of the change:

```
Dec 16 12:48:28 UA-KVM1 kernel: CPU1 has been hot-added
Dec 16 12:48:28 UA-KVM1 kernel: SMP alternatives: switching to SMP code
Dec 16 12:48:57 UA-KVM1 kernel: smpboot: Booting Node 0 Processor 1 APIC 0x1
Dec 16 12:48:57 UA-KVM1 kernel: kvm-clock: cpu 1, msr 0:3ff87041, secondary cpu clock
Dec 16 12:48:57 UA-KVM1 kernel: TSC synchronization [CPU#0 -> CPU#1]:
Dec 16 12:48:57 UA-KVM1 kernel: Measured 906183720569 cycles TSC warp between CPUs, turning off TSC clock.
Dec 16 12:48:57 UA-KVM1 kernel: tsc: Marking TSC unstable due to check_tsc_sync_source failed
Dec 16 12:48:57 UA-KVM1 kernel: KVM setup async PF for cpu 1
Dec 16 12:48:57 UA-KVM1 kernel: kvm-stealtime: cpu 1, msr 3fd0d240
Dec 16 12:48:57 UA-KVM1 kernel: microcode: CPU1 sig=0x206c1, pf=0x1, revision=0x1
Dec 16 12:48:57 UA-KVM1 kernel: Will online and init hotplugged CPU: 1
```

## Describe possible alternatives

We could give every firecracker microvm access to all cores and only use cgroups to limit actual scheduling time. This is not great though as it might create a lot of CPU steal. We prefer to give full cores when possible.

## Checks

- [x] Have you searched the Firecracker Issues database for similar requests?
- [x] Have you read all the existing relevant Firecracker documentation?
- [x] Have you read and understood Firecracker's core tenets?

**Comment by KarthikNedunchezhiyan** [comment]
@jeromegn does vertical scaling is the only solution to the problem you are to trying to fix? Does horizontal scaling like multiple instances not helped. Just curious to understand the usecase.

**Comment by raduiliescu** [comment]
Hi @jeromegn!

You might also want to look at cpu online/offline kernel feature - https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-devices-system-cpu.
On the downside you will always need to start the microVM with the max number of CPUs, and you will need an agent inside the microVM to write the values in /proc, but the overhead of keeping CPUs offline is better than in the cgroup case.

**Comment by jeromegn** [comment]
@KarthikNedunchezhiyan we need to support a large variety of workloads. More VMs isn't always the solution, but we're already doing that.

@raduiliescu thanks! That _could_ work, but our users have `root` access to the VM and could bring up any number of CPUs.

**Comment by AlexandruCihodaru** [comment]
We need to think about it for a bit, we will get back to you.

**Comment by sudanl0** [comment]
marking this as parked right now but we will track this as part of our roadmap.

**Comment by roypat** [comment]
Hey all, 
Just to leave a quick update here, we've prototyped vCPU hotplug in the [`feature/vcpu-hotplug` feature branch](https://github.com/firecracker-microvm/firecracker/tree/feature/vcpu-hotplug). The results of performance testing there was that the end-to-end latency of hotplugging and on-lining an addition vCPU was ~30ms per vCPU, which is prohibitively expensive for the post-restore scenario (where we want to be able to resume guest execution within single digit milliseconds). As such, we're moving this ticket back to the backlog for now :(

**Comment by wjordan** [comment]
Thanks for the update on this. Just to clarify, 30ms per vCPU sounds like an acceptable amount of latency for a vCPU hotplug process for most use cases, but it sounds like there's an issue in the prototype that requires re-adding previously-hotplugged vCPUs after a snapshot-restore, which is why this latency applies to the post-restore scenario.

In the linked branch, it sounds like the issue has something to do with the vCPU config not persisting after a restore:

https://github.com/firecracker-microvm/firecracker/blob/c41ceb5bd07528db0d8628b66a0fe7bbefc88114/docs/vcpu-hotplug.md?plain=1#L127-L132

Is this understanding correct? Does the next step in the implementation involve updating the snapshot state format to include the relevant vCPU hotplug info in the microVM state, or was there some other issue in the implementation that requires vCPU hotplug latency in the snapshot restore process?

**Comment by roypat** [comment]
Hi Will, 

> Just to clarify, 30ms per vCPU sounds like an acceptable amount of latency for a vCPU hotplug process for most use cases, but it sounds like there's an issue in the prototype that requires re-adding previously-hotplugged vCPUs after a snapshot-restore, which is why this latency applies to the post-restore scenario.

We only implemented pre-snapshot hotplugging back in July to get an idea of the latencies involved, but for the usecase we were looking at, we actually want to hotplug vCPUs only after restore (e.g. no hotplug ever happens before a snapshot is taken). 

> Is this understanding correct? Does the next step in the implementation involve updating the snapshot state format to include the relevant vCPU hotplug info in the microVM state, or was there some other issue in the implementation that requires vCPU hotplug latency in the snapshot restore process?

Essentially, the entire ACPI hotplug device would need to be persisted in the snapshot, but generally there shouldn't be much of an issue with that.

**Comment by aerosouund** [comment]
Hello @JamesC1305 @roypat @zulinx86 

I have a few follow up questions on where this feature is headed to know if i can support in helping it land to main.

Based on this:

>vCPU hot-plugging uses ACPI methods to enable the new CPUs inside the guest.
Currently, ACPI support is only available for x86_64 machines, and so it follows
that vCPU hot-plugging is only possible on x86_64 for the time being.

is it currently desired to complete the feature for aarch64 as well ? or is the initial release (merge to main) of this feature going to target x86_64 only ? can i support with adding the necessary changes to aarch64 as well ?

>(e.g. no hotplug ever happens before a snapshot is taken).

why ?

>Essentially, the entire ACPI hotplug device would need to be persisted in the snapshot, but generally there shouldn't be much of an issue with that.

what are you referring to ? the `CpuContainer` device that gets initiated in `attach_cpu_container_device` ?

>Hot-unplugging has not yet been implemented.

did you have any idea how that may look like at the API level ? maybe something like `PUT /hotplug -d {"Vcpu": {"del": 2}}` granted that the vm will have one or more CPUs remaining afterwards ?

**Comment by JamesC1305** [comment]
Hi @aerosouund,

Thank you for your interest in the implementation of vCPU hot-plugging. 

We would eventually like to have vCPU hot-(un)plug support for aarch64 as well as x86_64, but we have yet to decide how exactly that would look. Ideally, we would have one solution that works for both architectures, rather than maintaining per-arch implementations. This is not currently possible as we do not implement ACPI emulation for ARM.

Given the latency results we found from our prototype, we do not currently have a use case for vCPU hot-plugging, and have therefore de-prioritized its implementation. If there was a strong community demand and use-cases for hot-plugging, we'd likely have to discuss internally and come back with a design that we would be willing to support and maintain before accepting any contributions. 

Regarding your comments on the details of our previous implementation, it was rather minimal, non-exhaustive prototype just to be able to gather results regarding the latency of the CPU hot-plugging operation. If we wanted to expand upon that to implement hot-plugging for VMs booted from a snapshot, then yes, it would be the `CpuContainer` ACPI device that would need to be persisted. I did implement a hot-unplugging prototype (#4753), it just never got merged into our feature branch. If interested, you can see how I modified the API accordingly there.

**Comment by aerosouund** [comment]
@JamesC1305 
What i gather from your comment is that before committing to delivering a feature in that area, you need something that performs better and it requires time investment on the team's side that may be unavailable in the present. Let me know if i am mistaken.
In any case, is there anything i can contribute to in this area to push the needle forward ?
